[RNMobile] Ensure uploaded audio is always visible within Audio block

SiobhyB opened the pull request:
* Gutenberg Mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/6340

## What?

Fixes https://github.com/WordPress/gutenberg/issues/54328, by ensuring the Audio block is always visible when media's been uploaded to it.

## Why?

The content of an Audio block with set media is currently invisible when a block-based theme's background colour contrasts with the device's light/dark light mode. For example, if the theme's background colour is white and dark mode is enabled on a device, the Audio block's content will be invisible.

## How?

To account for a block based theme's background colour, a custom hook, `useEditorColorScheme`, has been created. The hook works as follows:

- Accepts both light and dark styles as its parameters.
- If no block-based theme is detected, `getStylesFromColorScheme()` is used to apply the appropriate style based on whether light or dark mode is enabled on a device. This is how the styles were previously applied. 
- If a block-based theme, its background colour is passed to the `colord.isDark()` function to determine if its dark or not. Then, the result is used to determine the appropriate colour scheme to use.

## Testing Instructions

Follow the _preparation steps_ once for each mode (dark/light) and then proceed with the _testing steps_.

### Preparation steps

* Ensure the device is in either dark mode or light mode.
* Activate a theme with support for global styles, e.g. [Byrne](https://byrnedemo.wordpress.com/).
* In your site's /wp-admin area, navigate to the site editor (Appearance → Editor) and select a style corresponding to the opposite mode (dark background for light mode, and vice versa).
* In the app, navigate to the editor for any of the site's posts or pages.

### Testing steps

* Tap the + icon to open the block inserter and add an Audio block.
* Follow the prompts to upload an audio file to the block.
* Verify that you're able to see the uploaded file within the block, along with prompts to play it.

### Additional testing

* Ensure there are no regressions associated with non-block based themes i.e. dark mode should still retain dark styles and vice versa.
* Experiment with non-conventional dark/light background colours with block based themes i.e. colours other than white or black.
* Test other changes styles, such as the failed upload styles.

### Testing Instructions for Keyboard

N/A, solely a visual change.

## Screenshots or screencast 

<details>
 <summary>Dark mode + block based theme with light background ⤵️</summary>
 
| Before  | After |
| ------------- | ------------- |
| <img src="https://github.com/WordPress/gutenberg/assets/2998162/98277a16-61b3-4dc2-8176-e9ae3e0a7000" width="100%"> | <img src="https://github.com/WordPress/gutenberg/assets/2998162/780e3408-687d-484a-a623-5820a8acb352" width="100%"> |
</details>

<details>
 <summary>Light mode + block based theme with dark background ⤵️</summary>
 
| Before  | After |
| ------------- | ------------- |
| <img src="https://github.com/WordPress/gutenberg/assets/2998162/71cc205d-d647-4562-a643-9f0decb22415" width="100%"> | <img src="https://github.com/WordPress/gutenberg/assets/2998162/6d43f8e4-185c-4fd6-ad63-9a4a01907022" width="100%"> |
</details>

<details>
<summary>No changes when dark/light mode matches background ⤵️</summary>
 
| Dark mode + dark background  | Light mode + light background |
| ------------- | ------------- |
| <img src="https://github.com/WordPress/gutenberg/assets/2998162/6ca03a7c-0833-4ce5-8d4a-82563409104b" width="100%"> | <img src="https://github.com/WordPress/gutenberg/assets/2998162/86ef5e25-7c75-405a-9dfc-79ad45cb44d4" width="100%"> |
</details>

<details>
 <summary>No changes for non-block based themes ⤵️</summary>
 
| Dark mode  | Light mode |
| ------------- | ------------- |
| <img src="https://github.com/WordPress/gutenberg/assets/2998162/d310ef07-7610-45cd-b5d4-ada9cfadc364" width="100%"> | <img src="https://github.com/WordPress/gutenberg/assets/2998162/7808e206-7857-4327-a741-23aadf06eef2" width="100%"> |
</details>

SiobhyB committed:
Update CHANGELOG

SiobhyB committed:
Update ticket reference in CHANGELOG

SiobhyB committed:
Refactor logic for dark/light mode styling

With this commit, the existing "getStylesFromColorScheme" function that applies different styles based on dark/light mode is removed. In its place, a custom function, "determineEditorColorScheme", has been added.

In future commits, we'll iterate on this groundwork and "determineEditorColorScheme" will begin to account for any editor styles that come from block-based themes.

SiobhyB committed:
Account for block-based theme's background color

To account for a block-based theme's background color, a custom hook has been created called "useEditorColorScheme". The hook works as follows:

- Accepts both the light and dark styles for each component.
- If there is no block-based theme, with a custom background color, the device's light/dark mode is used to determine the appropriate color scheme.
- If there is a block-based theme, the colord.isDark() function is used to test whether the background is dark or not. The result is used to determine the appropriate color scheme to use.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/audio-block-visibility

SiobhyB committed:
Fix positioning of CHANGELOG entry

fluiddot commented on packages/components/src/mobile/audio-player/index.native.js:
```diff
@@ -70,14 +71,30 @@ function Player( {
+		const editorColors = globalStyles?.baseColors?.color;
+		const editorBackgroundColor = editorColors?.background;
+
+		if ( typeof editorBackgroundColor === 'undefined' ) {
+			return getStylesFromColorScheme( baseStyle, darkStyle );
+		}
```
I noticed that `editorBackgroundColor` value can also be a String with the value `undefined` when the color can't be parsed:

https://github.com/WordPress/gutenberg/blob/531089c7e2f3c34aaf3b782a1a9aa579f369be0a/packages/components/src/mobile/global-styles-context/utils.native.js#L34)
https://github.com/WordPress/gutenberg/blob/531089c7e2f3c34aaf3b782a1a9aa579f369be0a/packages/components/src/mobile/global-styles-context/utils.native.js#L226-L236
https://github.com/WordPress/gutenberg/blob/531089c7e2f3c34aaf3b782a1a9aa579f369be0a/packages/components/src/mobile/global-styles-context/utils.native.js#L260-L268

You can check this by using the theme `Twenty Twenty-Three`. In this case, the Audio block is not visible.

<img src=https://github.com/WordPress/gutenberg/assets/14905380/99827860-ae85-4f0d-90c0-070ad5f3c447 width=200>

fluiddot replied:
For some reason, I only reproduce the issue I mentioned above when using the iOS simulator. On an actual device, themes and colors are parsed properly 🙃. In any case, adding the condition for the case I mentioned, when values can't be parsed, would still be applicable.

dcalhoun replied:
Noting that I also experienced an illegible Audio block when using the Twenty Twenty-Three theme and dark mode enable **only when using a simulator**. When compared to a physical device, the editor's background color is different. 


| Simulator | Physical Device | 
| - | - |
| <img alt="Illegible Audio block details with Twenty Twenty-Three theme and dark mode on simulator" src="https://github.com/WordPress/gutenberg/assets/438664/7e45f9cf-6a10-42e9-8cab-570ba8d4708f" /> | ![Legible Audio block details with Twenty Twenty-Three theme and dark mode on physical iPhone SE](https://github.com/WordPress/gutenberg/assets/438664/4f8d5587-c3e0-4c30-83fb-6b2f6d84a2aa) |

SiobhyB replied:
@fluiddot, @dcalhoun, thank you both! It's strange that this is specific to iOS (not Android) and only reproducible on a simulator. 🤔 I also had trouble reproducing myself, but did not that the editor's background isn't always being displayed on an iOS simulator for me. It feels like there's something more to be dug into here.

Regardless, I've added a check for the "undefined" string, to resolve this case from the code Carlos pointed out: https://github.com/WordPress/gutenberg/pull/55627/commits/e83f5c688c2b46c66d5559d1f373419aa64ef2e5 (also viewable in the refactored function in https://github.com/WordPress/gutenberg/pull/55627/commits/190bbd07eb174150b09f0beb1da04404c66030ba).

Let me know if that resolves this issue for you!

fluiddot commented on packages/components/src/mobile/audio-player/index.native.js:
```diff
@@ -70,14 +71,30 @@ function Player( {
 	};
 
-	const containerStyle = getStylesFromColorScheme(
+	const useEditorColorScheme = ( baseStyle, darkStyle ) => {
```
`useEditorColorScheme` provides functionality that I presume might be needed on other blocks, as choosing styles based on the background seems a common issue across blocks. I wonder if we should move it as a utility related to global styles. WDYT?

SiobhyB replied:
That's a good point! I've gone ahead to refactor the function in https://github.com/WordPress/gutenberg/pull/55627/commits/190bbd07eb174150b09f0beb1da04404c66030ba, with that reusability in mind.

I also wanted to highlight that this _might_ be a short-term patch, as I plan to re-open the discussion of how we handle block-based/dark-mode colours in the long-term with the team.

dcalhoun commented on packages/react-native-editor/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ For each user feature we should also add a importance categorization label  to i
 -->
 
 ## Unreleased
+-   [*] Audio block: Fix visibility of upload audio [#55627]
```
Given we use this to source user-facing notes, it may be beneficial to capture more detail. WDYT? 
```suggestion
-   [*] Audio block: Improve legibility of audio file details on various background colors [#55627]
```

SiobhyB replied:
The suggested change looks great, thanks @dcalhoun! I update the note in Gutenberg Mobile too, in https://github.com/wordpress-mobile/gutenberg-mobile/pull/6340/commits/15f16aacec1531fbbaa56d0721fc238e7aacc9db.

SiobhyB committed:
Iterate on CHANGELOG description

Co-authored-by: David Calhoun <github@davidcalhoun.me>

SiobhyB committed:
Merge branch 'trunk' into rnmobile/audio-block-visibility

SiobhyB committed:
Fallback to device style if theme styles undefined

As described in the following feedback, there can be times when the editor background colour returns an "undefined" string: https://github.com/WordPress/gutenberg/pull/55627/#discussion_r1377462299

With this commit, that case is accounted for, with the default dark/light styles being applied.

SiobhyB committed:
Refactor 'useEditorColorScheme' for usability

As per the feedback here, the function may be useful for other instances beyond the audio block, so it has been refactored and moved to a utils file: https://github.com/WordPress/gutenberg/pull/55627#discussion_r1377467459

Note, to comply with the rules of hooks, it was also necessary to reorganise the functions logic as part of this refactor:

- The "usePreferredColorScheme" replaced "getStylesFromColorScheme" as the function for detecting the device's light/dark mode.
- "usePreferredColorScheme" is called after first checking to see if the editor's background is defined.

SiobhyB commented:
Noting failing unit tests related to the audio block following the last commits, I'm looking into them and will set the PR back as ready for review when I've addressed them.

SiobhyB committed:
Call correct hook for returning color scheme style

"usePreferredColorScheme" was previously used to return whether the device was in dark or light mode but, following a refactor, "usePreferredColorSchemeStyle" is now needed. This is because we're now returning style objects rather than simply the dark/light mode string.

SiobhyB commented:
Returning to this with fresh eyes, after it fell off my list of priorities, I spotted the bug that was leading to failing tests and fixed it in https://github.com/WordPress/gutenberg/pull/55627/commits/51490f00f252388cd196a5c1d6d460e6be34e050. This PR's ready for review again. :)

fluiddot commented on packages/components/src/mobile/global-styles-context/utils.native.js:
```diff
@@ -436,3 +443,40 @@ export function getGlobalStyles( rawStyles, rawFeatures ) {
+ *
+ * @return {Object} - The combined style object that should be applied to the editor.
+ */
+export const useEditorColorScheme = ( baseStyle, darkStyle ) => {
```
Not a blocker for this PR but eventually it would be great to have unit tests to cover this hook.

SiobhyB replied:
Thanks Carlos, I'll add this to my backlog.

fluiddot reviewed (approved):
LGTM 🎊 ! I've tested using different themes and worked as expected, awesome work @SiobhyB 🏅 !

fluiddot commented:
@SiobhyB I encountered several React warnings when opening the editor using these changes. However, they were solved after updating with the latest changes from `trunk`. Sharing this in case someone else experiences it when testing the PR.

SiobhyB committed:
Merge branch 'trunk' into rnmobile/audio-block-visibility

fluiddot commented:
@SiobhyB FYI I managed to reproduce the issue related to this PR in version `23.8` (https://github.com/WordPress/gutenberg/issues/54328#issuecomment-1829937617).
